contracts/LockedNORIV2.sol

```diff
@@ -0,0 +1,788 @@
+    uint256 vestEndTime;
+    uint256 unlockEndTime;
+    uint256 cliff1Time;
+    uint256 cliff2Time;
+    uint256 vestCliff1Amount;
+    uint256 vestCliff2Amount;
+    uint256 unlockCliff1Amount;
+    uint256 unlockCliff2Amount;
+  }
+
+  struct DepositForParams {
+    address recipient;
+    uint256 startTime;
+  }
+
+  /**
+   * @notice Role conferring creation and revocation of token grants.
+   */
+  bytes32 public constant TOKEN_GRANTER_ROLE = keccak256("TOKEN_GRANTER_ROLE");
+
+  /**
+   * @notice Used to register the ERC777TokensRecipient recipient interface in the
+   * ERC-1820 registry
+   *
+   * @dev Registering that LockedNORI implements the ERC777TokensRecipient interface with the registry is a
+   * requiremnt to be able to receive ERC-777 BridgedPolygonNORI tokens. Once registered, sending BridgedPolygonNORI
+   * tokens to this contract will trigger tokensReceived as part of the lifecycle of the BridgedPolygonNORI transaction
+   */
+  bytes32 public constant ERC777_TOKENS_RECIPIENT_HASH =
+    keccak256("ERC777TokensRecipient");
+
+  /**
+   * @notice A mapping from grantee to grant
+   */
+  mapping(address => TokenGrant) private _grants;
+
+  /**
+   * @notice The BridgedPolygonNORI contract that this contract wraps tokens for
+   */
+  IERC20Upgradeable private _bridgedPolygonNori;
+
+  /**
+   * @notice The [ERC-1820](https://eips.ethereum.org/EIPS/eip-1820) pseudo-introspection registry
+   * contract
+   *
+   * @dev Registering that LockedNORI implements the ERC777TokensRecipient interface with the registry is a
+   * requiremnt to be able to receive ERC-777 BridgedPolygonNORI tokens. Once registered, sending BridgedPolygonNORI
+   * tokens to this contract will trigger tokensReceived as part of the lifecycle of the BridgedPolygonNORI transaction
+   */
+  IERC1820RegistryUpgradeable private _erc1820;
+
+  /**
+   * @notice Emitted on successful creation of a new grant.
+   */
+  event TokenGrantCreated(
+    address indexed recipient,
+    uint256 indexed amount,
+    uint256 indexed startTime,
+    uint256 vestEndTime,
+    uint256 unlockEndTime
+  );
+
+  /**
+   * @notice Emitted on when the vesting portion of an active grant is terminated.
+   */
+  event UnvestedTokensRevoked(
+    uint256 indexed atTime,
+    address indexed from,
+    uint256 indexed quantity
+  );
+
+  /**
+   * @notice Emitted on withdwal of fully unlocked tokens.
+   */
+  event TokensClaimed(
+    address indexed from,
+    address indexed to,
+    uint256 quantity
+  );
+
+  /**
+   * @notice Emitted when the underlying token contract address is updated due to migration.
+   */
+  event UnderlyingTokenAddressUpdated(address from, address to);
+
+  /**
+   * @notice This function is triggered when BridgedPolygonNORI is sent to this contract
+   *
+   * @dev Sending BridgedPolygonNORI to this contract triggers the tokensReceived hook defined by the ERC-777 standard
+   * because this contract is a registered ERC777 tokens recipient.
```

Comment:
could probably make this language less specific now?

Reply:
Good call I need to go back and look at these docs.